src/oauth/sessionManager.ts

```diff
@@ -0,0 +1,808 @@
+			grant_type: AUTH_GRANT_TYPE,
+			code,
+			redirect_uri: this.getRedirectUri(),
+			client_id: registration.client_id,
+			client_secret: registration.client_secret,
+			code_verifier: verifier,
+		};
+
+		const tokenRequest = toUrlSearchParams(params);
+
+		const response = await axiosInstance.post<TokenResponse>(
+			metadata.token_endpoint,
+			tokenRequest,
+			{
+				headers: {
+					"Content-Type": "application/x-www-form-urlencoded",
+				},
+			},
+		);
+
+		this.logger.info("Token exchange successful");
+
+		await this.saveTokens(response.data);
+
+		return response.data;
+	}
+
+	/**
+	 * Refresh the access token using the stored refresh token.
+	 * Uses a shared promise to handle concurrent refresh attempts.
+	 */
+	public async refreshToken(): Promise<TokenResponse> {
+		// If a refresh is already in progress, return the existing promise
+		if (this.refreshPromise) {
+			this.logger.debug(
+				"Token refresh already in progress, waiting for result",
+			);
+			return this.refreshPromise;
+		}
+
+		if (!this.storedTokens?.refresh_token) {
+			throw new Error("No refresh token available");
+		}
+
+		const refreshToken = this.storedTokens.refresh_token;
+		const accessToken = this.storedTokens.access_token;
+
+		this.lastRefreshAttempt = Date.now();
+
+		// Create and store the refresh promise
+		this.refreshPromise = (async () => {
+			try {
+				const { axiosInstance, metadata, registration } =
+					await this.prepareOAuthOperation(accessToken);
+
+				this.logger.debug("Refreshing access token");
+
+				const params: RefreshTokenRequestParams = {
+					grant_type: REFRESH_GRANT_TYPE,
+					refresh_token: refreshToken,
+					client_id: registration.client_id,
+					client_secret: registration.client_secret,
+				};
+
+				const tokenRequest = toUrlSearchParams(params);
+
+				const response = await axiosInstance.post<TokenResponse>(
+					metadata.token_endpoint,
+					tokenRequest,
+					{
+						headers: {
+							"Content-Type": "application/x-www-form-urlencoded",
+						},
+					},
+				);
+
+				this.logger.debug("Token refresh successful");
+
+				await this.saveTokens(response.data);
+
+				return response.data;
+			} finally {
+				this.refreshPromise = null;
+			}
+		})();
+
+		return this.refreshPromise;
+	}
+
+	/**
+	 * Save token response to storage.
+	 * Also triggers event via secretsManager to update global client.
+	 */
+	private async saveTokens(tokenResponse: TokenResponse): Promise<void> {
+		const deployment = this.requireDeployment();
+		const expiryTimestamp = tokenResponse.expires_in
+			? Date.now() + tokenResponse.expires_in * 1000
+			: Date.now() + ACCESS_TOKEN_DEFAULT_EXPIRY_MS;
+
+		const tokens: StoredOAuthTokens = {
+			...tokenResponse,
+			deployment_url: deployment.url,
+			expiry_timestamp: expiryTimestamp,
+		};
+
+		this.storedTokens = tokens;
+		await this.secretsManager.setOAuthTokens(deployment.label, tokens);
+		await this.secretsManager.setSessionAuth(deployment.label, {
```

Comment:
From the outside looking in, it feels weird to me that we set oauth tokens and session auth separately.  I think I would expect a single session auth object, and that object would have oauth information as well if it came from oauth.

Or to approach this another way, is there a use case for setting only oauth tokens but not the session auth?